src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,500 @@
+   * @param pos position of the flag to test, from the most significant bit the the least significant bit.
+   * @return the value of the flag.
+   */
+  bool GetFlag(uint32_t pos) { return Bitmap().Test(pos); }
+
+  void SetVersion(uint32_t version) { version_ = version; }
+
+  uint32_t GetVersion() { return version_; }
+
+ private:
+  uint32_t checksum_;
+  // Note: use uint32_t instead of table_oid_t, because table_oid_t is not POD so that
+  // we cannot use PACKED in this class.
+  uint32_t table_oid_;
+  uint32_t version_;  // version of the schema(useful with schema change), or, for catalog tables,
+                      // serves as the type of the catalog table.
+  /* 32 bit flag. The meaning of each bit is:
+   *  0. catalog table
+   */
+  uint32_t flags_;
+  byte payload_[0];
+
+  common::RawBitmap &Bitmap() { return *reinterpret_cast<common::RawBitmap *>(&flags_); }
+
+  const common::RawBitmap &Bitmap() const { return *reinterpret_cast<const common::RawBitmap *>(&flags_); }
+};
+
+/**
+ * An async block writer will use a background thread to write blocks into
+ * checkpoint file asynchronously. Its design is very similar with that of
+ * the logger threads in SiloR. An async block writer will have some
+ * pre-allocated block buffers. Workers must first ask for a free block
+ * buffer. If there is no free block buffer now, it will be blocked until
+ * some free block buffer is available. In this way, the async block writer
+ * can give some back pressure on the worker to let it slow down if it is
+ * generating blocks too quickly. The worker then fills the block buffer
+ * and hand it back to the writer thread. The writer thread writes it into
+ * the file and make the buffer available again.
+ */
+class AsyncBlockWriter {
+ public:
+  AsyncBlockWriter() = default;
+
+  /**
+   * Instantiate a new AsyncBlockWriter, open a new checkpoint file to write into.
+   * @param log_file_path path to the checkpoint file.
+   * @param buffer_num number of pre-allocated buffer
+   */
+  explicit AsyncBlockWriter(const char *log_file_path, int buffer_num) { Open(log_file_path, buffer_num); }
+
+  /**
+   * Open a file to write into
+   * @param log_file_path path to the checkpoint file.
+   * @param buffer_num number of pre-allocated buffer
+   */
+  void Open(const char *log_file_path, int buffer_num);
+
+  /**
+   * Ask for a free block buffer
+   * @return pointer to a free block buffer
+   */
+  byte *GetBuffer() {
+    byte *res = nullptr;
+    free_.Dequeue(&res);
+    return res;
+  }
+
+  /**
+   * Let the writer write a buffer (async)
+   * @param buf the pointer to the buffer to be written
+   */
+  void WriteBuffer(byte *buf) {
+    TERRIER_ASSERT(buf != nullptr, "Buffer pointer should not be nullptr");
+    pending_.Enqueue(buf);
+  }
+
+  /**
+   * Return a buffer to the writer, make it available again
+   * without writing it
+   * @param buf the pointer to the buffer to be returned
+   */
+  void ReturnBuffer(byte *buf) {
```

Comment:
Should this be exposed in the interface? In what situations would someone need to GetBuffer and return it without calling WriteBuffer. I only see it being used in BufferedTupleWriter.close, which also calls AsyncBlockWriter.close(), so why no just clean up the free queue in that function?

Reply:
AsyncBlockWriter allocates these buffers. Therefore, these buffers had better be freed by the AsyncBlockWriter itself. Requiring BufferedTupleWriter to free these buffers leaks the detail of buffer management out of the AsyncBlockWriter abstraction, which makes it harder to ensure the lifetime of these buffers is correct. I think it is not the right thing to do.